gix-worktree-state/tests/state/checkout.rs

```diff
@@ -291,6 +319,54 @@ fn dangling_symlinks_can_be_created() -> crate::Result {
     Ok(())
 }
 
+#[test]
+fn dangling_symlink_to_windows_invalid_target_can_be_created() -> crate::Result {
```

Comment:
Could you combine the two test fixtures and the two tests itself into one?

They seem to be copy-paste with exactly one part changed (the target of the symlink), and in that case I think it's better to consolidate.

The fixture can reasonably create two dangling links at once as well.

If you prefer to keep the fixtures separate to make them more descriptive, I am up for that too, but truly believe the Rust side should be consolidated.

You will find the following useful.

```Rust
for (fixture, expected_name) in [("fixture1", "CON"), ("fixture2", "???")] {

}
```

Reply:
Yes, absolutely. Unfortunately I have to go away from my computer for a while right now, but I plan to do this when I get back.

I can see how it is more important to consolidate the Rust code of the tests than the fixtures, since among other things the Rust code of the tests is significantly longer and more complicated. I'll try to figure out which of the two approaches is better--in terms of whether the fixtures should also be consolidated--when I return.

The tests are based very heavily on the `dangling_symlinks_can_be_created` test, and maybe all three could even be combined into one. Then again, these tests, `dangling_symlink_to_windows_invalid_target_can_be_created` and `dangling_symlink_to_windows_reserved_target_can_be_created`, are somewhat more conceptually related to each other than they are to that test. If you have a preference as to whether all three could reasonably be combined, you can let me know. Otherwise, I will also try to figure that out when I come back.

The `symlinks_to_directories_are_usable` test is also somewhat similar, though less so, because it has an additional assertion--the most important of the assertions in it--that the metadata are accessible through the symlink.

> You will find the following useful.

Yes, this pattern appears throughout the test suite and I should really have just used it originally based on that.

I am curious if there is a reasonably uncomplicated way to parameterize test functions, whether through the testing framework, or by using the genericity features of the Rust language itself, so that they are run as independent tests with independent statuses, yet still without code duplication. (Don't worry; that curiosity will not prevent me from using this pattern, which is a decisive improvement compared to what I currently have written, and which can always be improved on further if such a thing could/should ultimately be done.)

Reply:
>  If you have a preference as to whether all three could reasonably be combined, you can let me know. Otherwise, I will also try to figure that out when I come back.

I am also very cautious about trying to deduplicate too much - if it's low-hanging, then it should be done, but to my mind there is no need to consolidate more unless one feels it's a problem.

> The `symlinks_to_directories_are_usable` test is also somewhat similar, though less so, because it has an additional assertion--the most important of the assertions in it--that the metadata are accessible through the symlink.

I thought it's specific enough to not try to consolidate. In the test-suite, there are a few examples for complex test-programs with lots of parameters effectively, and despite having good use, they are also hard to understand. It's always a trade-off, but generally I don't spend much brainpower on deduplication, but do it if it's very obvious like here.

> I am curious if there is a reasonably uncomplicated way to parameterize test functions, whether through the testing framework, or by using the genericity features of the Rust language itself, so that they are run as independent tests with independent statuses, yet still without code duplication. (Don't worry; that curiosity will not prevent me from using this pattern, which is a decisive improvement compared to what I currently have written, and which can always be improved on further if such a thing could/should ultimately be done.)

I think the answer is 'yes', but with the right framework. There are property-test crates out there that provide proc-macros to annotate functions, even though I have not used them.

Once somebody introduced a new test-runner that allows to procedurally create test-objects with closures, but ultimately I removed it as it was harder to deal with while some of the tests aren't passing. [This is the file I am talking about, it's history should reveal everything](
https://github.com/Byron/gitoxide/blob/3ef3bc20a1b90799e5ac26858f898bc7a7c96901/gix-url/tests/baseline.rs#L13).

Custom test-runners are not worth it here, but certain utility crates might be, I just didn't use them yet.
Personally, I prefer the simple loop if there isn't too much data, and baseline programs if there is more. Direct Rust code, no magic.

Reply:
I ended up consolidating all three dangling symlink tests rather than just the two new ones, though I'd be pleased to consolidate just the two new ones if you prefer that.

The reason I ended up deciding to consolidate all three is that it shows the similarities between what the are testing. Keeping them fully separate had obscured that somewhat; it seems to me that consolidating two of them but not the other might obscure the relationship much more.

Part of this reflects my belief that #1420 should be understood primarily as being about cases of dangling symlinks that were not covered before (per the "A note on the semantics of such symlinks" subsection there). However, that is was actually less of a factor in my thinking than the value that it seems to me comes from seeing exactly what varies across all three of those tests:

https://github.com/Byron/gitoxide/blob/87422ab34922131e3243d5d7db7910a3f68978a4/gix-worktree-state/tests/state/checkout.rs#L306-L310

However, to reiterate, I am fully willing to split the original test and the two new ones, so that only the two new ones are deduplicated.

I have of course, and as mentioned above, *not* attempted to deduplicate anything to do with the new `symlinks_to_directories_are_usable` test, which remains separate from all others, new and old.